src/Cascade/Cascade.man
.\" t
.\" @(#)Cascade.1  02/07/00
.TH Cascade 1.8 "Feb 7 2000" Cascade
.UC
.SH NAME
\fBCascade\fP \- the AfterStep window cascading module

.SH SYNOPSIS
\fBCascade\fP [\-f \fIconfig_file\fP] [\-v | \--version] [\-h | \--help]
[\-a] [\-desk] [\-flatx] [\-flaty] [\-incx \fIarg\fP] [\-incy \fIarg\fP] 
[\-m] [\-noraise] [\-nostretch] [\-r] [\-resize] [\-s] [\-t] [\-u] 
[x y [width height]]

.sp
Although \fBCascade\fP can be run from the command line, it is normally 
spawned by afterstep in the \fIautoexec\fP file, in the following manner:
.nf
.sp
% cat "~/GNUstep/Library/AfterStep/start/Modules/Cascade"
	Module "Animate" Cascade -resize 10 2 80 70
	MiniPixmap "mini-as.xpm"
.fi

To stop the \fBCascade\fP module, a menu item can be used:
.nf
.sp
% cat "~/GNUstep/Library/AfterStep/start/Modules/Cascade_Stop"
	KillModuleByName "Stop Cascade" Cascade
	MiniPixmap "mini-as.xpm"
.fi

.SH DESCRIPTION
The \fBCascade\fP module attempts to cascade windows on the current screen
subject to certain constraints.  Layering is performed so consecutive
windows will have their window titles visible underneath the previous.

.SH OPTIONS
.IP "\-f \fIconfig_file\fP"
Use \fIconfig_file\fP instead of the default configuration file.

.IP "\-h | \--help"
Print a usage message and exit successfully.

.IP "\-v | \--version"
Print version information and exit successfully.

.IP "\-a"
Causes \fIall\fP window styles to be affected, even ones with the
WindowListSkip style.

.IP "\-desk"
Causes all windows on the desk to be cascaded instead of the current
screen only.

.IP "\-flatx"
Inhibits border width increment.

.IP "\-flaty"
Inhibits border height increment.

.IP "\-incx \fIarg\fP"
Specifies a horizontal increment which is successively added to
cascaded windows.  \fIarg\fP is a percentage of screen width, or pixel
value if a \fIp\fP is suffixed.  Default is zero.

.IP "\-incy \fIarg\fP"
Specifies a vertical increment which is successively added to cascaded
windows.  \fIarg\fP is a percentage of screen height, or pixel value
if a \fIp\fP is suffixed.  Default is zero.

.IP "\-m"
Causes maximized windows to also be affected (implied by \-all).

.IP "\-noraise"
Inhibits window raising, leaving the depth ordering intact.

.IP "\-nostretch"
Inhibits window expansion when using the \-resize option.  Windows
will only shrink to fit the maximal width and height (if given).

.IP "\-r"
Reverses the window sequence.

.IP "\-resize"
Forces all windows to resize to the constrained width and height (if
given).

.IP "\-s"
Causes sticky windows to also be affected (implied by \-all).

.IP "\-t"
Causes transient windows to also be affected (implied by \-all).

.IP "\-u"
Causes untitled windows to also be affected (implied by \-all).

.SH INVOCATION
\fBCascade\fP can be invoked from binding the action "Module Cascade
\fIoptions\fP" to a key-stroke in the \fIfeel\fP, or from a menu item as
shown in the Synopsis section above.

The \fBCascade\fP module can also be called multiple times with
different options. This way multiple Cascading programs can be used, and
is unlike other modules which require a symlink to run multiple times.

.SH CONFIGURATION OPTIONS
Command-line arguments passed to \fBCascade\fP are described in the 
\fIOPTIONS\fP section, above.

Up to four numbers can be placed on the command line that are not
switches.  The first pair specify an x and y offset to start the first
window (default is 0, 0).  The second pair specify a maximal width and
height for the layered windows.  If an affected window exceeds either
this width or height, it is resized to the maximal width or height.
If each number is suffixed with the letter p, then it is taken to be a
pixel value, otherwise it is interpreted as a screen percentage.
Specifying zero for any parameter is equivalent to not specifying it.

.SH EXAMPLE
As an example case, one could call Cascade with the following arguments:

\fBCascade \-resize 10 2 80 70\fP

This invocation will cascade windows starting 10 by 2 percent into and
down the screen.  Windows will be constrained to 80 by 70 percent of
the screen dimensions.  Since the \fIresize\fP is also specified,
windows will be resized to the given constrained width and height.

.SH BUGS
It is probably not a good idea to delete windows while windows are
being cascaded.

.SH AUTHOR
Andrew Veliath
.nf
David Mihm <dmihm@swbell.net> [Man page]

.SH SEE ALSO
.BR afterstep (1)


ChangeLog

2003-09-25  sasha

	* generating ChangeLog

2003-09-24  sasha

	* Completed look.Glass ; renamed PointerFore, PointerBack to
	  CursorFore, CursorBack to match old Cursor option; Moved Cursor
	  definitions from feel into look Completed Colorscheme's support
	  for Cursor's color

2003-09-23  sasha

	* Added look options for mouse pointer coloring: PointerFore and
	  PointerBack; Repaired code to update window's decorations when
	  stickiness flag is toggled;

2003-09-22  sasha

	* Added cursors to default feel file; enable cursor switching while
	  hovering over different elements of the interface; Added code to
	  change color of the cursor according to what colorscheme is set
	  to - may need to tweak a bit , or make colors configurable
	* Fixed locale parsing from LANG variable; Fixed module startup to
	  pass on current locae specified on command line

2003-09-19  sasha

	* added debugging code for cursor loading

2003-09-16  sasha

	* fixed bug with windows teleporting back to original position, if
	  moved/resized too fast

2003-09-15  sasha

	* updated configure scripts
	* Fixed configure.in to properly enable i18n in libAfterImage as
	  requested. Internationalized text should be displayed properly
	  now, even if its not in UTF8; Update src/test to work with new
	  libs and added test code for i18n testing
	* Fixed unquoted string in i18N code reported by Alex Mezin

2003-09-14  sasha

	* Added checks for locale and proper initialization of
	  libAfterImage t3ext drawing in accordance to its value; Fixed
	  problem with EdgeScroll being mangled when just look changes;
	  Fixed recently introduced bug causing wierd corners in framned
	  windows upon the look change
	* Fixed typoe for msg type in SendString being declared as some
	  wierd type of unsigned long long  - possibly causing wierdness on
	  64 bit machines

2003-09-13  sasha

	* implemented feedback functionality via sending the e-mail to
	  sasha :)

2003-09-12  sasha

	* added configure option to send e-mail back to developer on
	  installation; Fixed canvas border width handling while resizing
	  and in Pager;

2003-09-11  sasha

	* fixed handling of WM_PROTOCOL changes where WM_COMMAND would get
	  lost and session management would stop working; Found bug with
	  huge impact - all the geometry calculations were off as we were
	  forgetting border width of the frame window; Fixed handling of
	  button presses on border of the frame

2003-09-10  sasha

	* fixed NeXTish colorscheme somewhat
	* Fixed workspace state save/restore to allow terms to exec command
	  using -e, and to have argv's with spaces in them; Fixed frame
	  corner sizing to make it same width as is tbar size;

2003-09-09  sasha

	* More fixing Pager's geometry; Added -i cmd line parameter to
	  ascompose to allow for 'include' files (such as colorschemes,
	  etc.); Updated configure scripts to properly detect headers when
	  its run from libAfterBase; Updated version to beta2; Updated TEAM
	  file

2003-09-08  sasha

	* Fixed Pager to properly account for label size while sizing
	  min-desktop windows

2003-09-07  sasha

	* somewhat debugged Pager's geometry handling - need to add some
	  code to properly recalculate Pager's geometry if its soecified as
	  ratio of the main desktop vs. exact geometry

2003-09-03  sasha

	* added new AfterStep log icon; fixed wharf config file

2003-09-02  sasha

	* added new xml tag : <color> to define symbolic color names
	* changed handling of tiled images to clip it properly when
	  Right/Left align is used

2003-08-30  sasha

	* debugged workspace state handling code, and improved Wait loop

2003-08-29  sasha

	* mostly implemented new workspace state saving and restoring

2003-08-28  sasha

	* fixed workspace status generator to not save xrm params, as noone
	  seems to support thoise
	* added caching of pixmaps for unfocused bas in window decorations
	  to speed up fast focus switching
	* 64-bit fixes: CARD32 instead of Window type
	* 64-bit fixes: properties and strlen warnings
	* 64-bit fixes: converted intermodule communications to always use
	  32bit values
	* more 64 bit fixes

2003-08-27  sasha

	* 64-bit compilation fixes
	* Fixed slowness in intermodule communications manifesting itself
	  when desktops are switched
	* fixed handling of non-default MyBackgrounds, and proper adding of
	  DeskConfigs while switching desks
	* synching
	*  fixed segfaulting in Pager on look switches or after window
	  restacking
	* Fixed handling of hints flags - it was doing nothing before if
	  flags == 0 - was screwing MOtif Hints like totatlly; Fixed shaped
	  window handling in WinList; Added check to not do too much
	  redrawing when WinLIst is receiving window list - may want to do
	  the same in Pager

2003-08-26  sasha

	*  added parameters to import funcions, to allow for sized
	  image/preview stuff; Removed compression buffer from ASImage
	  structure to save up some memory; Hopefully fixed hints handling
	  on 64 bit hw

2003-08-24  sasha

	* fixed event handling when windows is getting mapped - no more
	  lost events due to not selected mask ; Fixed initial status when
	  mouse pointer is out of the screen; Fixed ICCCM size hints
	  handling to have BaseSize and MinSize interchange each other when
	  one is ommited - fixes GIMPs initial placement

2003-08-23  sasha

	* added new 3 icons to complete the set of default icons: switched
	  AfterStep to use ~/.afterstep instead of ridiculous ~/GNUstel/L/A
	  stuff

2003-08-22  sasha

	* Fixed handling of shaped windows - shape is not erroneously
	  cleared, and client's shape is subtracted from decoration's shape
	* fixed autodetection of zipped xpms and added 6 more icons to
	  default iconset

2003-08-21  sasha

	* more improvements to asimbrowser

2003-08-20  sasha

	* fixed makefile for ascp, when fltk is not found; added better
	  Style for BitchX to database; remnoved scrollbar from BitchX in
	  Wharf

src/Cascade/Cascade.c
/*
 * Copyright (c) 1997 Guylhem Aznar <guylhem@oeil.qc.ca>
 * Copyright (c) 1996 Andrew Veliath
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <sys/time.h>
#include <sys/types.h>
#include <unistd.h>
#include <signal.h>

#ifdef ISC
#include <sys/bsdtypes.h>
#endif

#if defined ___AIX || defined _AIX || defined __QNX__ || defined ___AIXV3 || defined AIXV3 || defined _SEQUENT_
#include <sys/select.h>
#endif

#include <X11/Xlib.h>
#include <X11/Xutil.h>

#define MODULE_X_INTERFACE

#include "../../configure.h"
#include "../../include/aftersteplib.h"
#include "../../include/afterstep.h"
#include "../../include/style.h"
#include "../../include/screen.h"
#include "../../include/module.h"

typedef struct window_item
  {
    Window frame;
    int th, bw;
    unsigned long width, height;
    struct window_item *prev, *next;
  }
window_item, *window_list;

/* vars */
Display *dpy;
int screen;
ScreenInfo Scr;
char *MyName = NULL;

static int fd_width;
static int fd[2];
static int x_fd;

window_list wins = NULL, wins_tail = NULL;
int wins_count = 0;

/* switches */
int ofsx = 0, ofsy = 0;
int maxw = 0, maxh = 0;
int untitled = 0, transients = 0;
int maximized = 0;
int all = 0;
int desk = 0;
int reversed = 0, raise_window = 1;
int resize = 0, nostretch = 0, sticky = 0;
int flatx = 0, flaty = 0;
int incx = 0, incy = 0;

/* masks for AS pipe */
#define mask_reg (	M_CONFIGURE_WINDOW |	\
			M_END_WINDOWLIST	)

void
insert_window_list (window_list * wl, window_item * i)
{
  if (*wl)
    {
      if ((i->prev = (*wl)->prev))
	i->prev->next = i;
      i->next = *wl;
      (*wl)->prev = i;
    }
  else
    i->next = i->prev = NULL;
  *wl = i;
}

void
free_window_list (window_list * wl)
{
  window_item *q;
  while (*wl)
    {
      q = (*wl)->next;
      free (*wl);
      *wl = q;
    }
}

int
is_suitable_window (unsigned long *body)
{
  XWindowAttributes xwa;
  unsigned long flags = body[8];

  if ((flags & WINDOWLISTSKIP) && !all)
    return 0;

  if ((flags & MAXIMIZED) && !maximized)
    return 0;

  if ((flags & STICKY) && !sticky)
    return 0;

  if (!XGetWindowAttributes (dpy, (Window) body[1], &xwa))
    return 0;

  if (xwa.map_state != IsViewable)
    return 0;

  if (!(flags & MAPPED))
    return 0;

  if (flags & ICONIFIED)
    return 0;

  if (!desk)
    {
      int x = (int) body[3], y = (int) body[4];
      int w = (int) body[5], h = (int) body[6];
      if (!((x < Scr.MyDisplayWidth) && (y < Scr.MyDisplayHeight)
	    && (x + w > 0) && (y + h > 0)))
	return 0;
    }
  if (!(flags & TITLE) && !untitled)
    return 0;

  if ((flags & TRANSIENT) && !transients)
    return 0;

  return 1;
}

int
get_window (void)
{
  ASMessage *asmsg = NULL;
  int last = 0;
  asmsg = CheckASMessage (fd[1], -1);
  if (asmsg)
    {
      switch (asmsg->header[1])
	{
	case M_CONFIGURE_WINDOW:
	  if (is_suitable_window (asmsg->body))
	    {
	      window_item *wi =
	      (window_item *) safemalloc (sizeof (window_item));
	      wi->frame = (Window) asmsg->body[1];
	      wi->th = (int) asmsg->body[9];
	      wi->bw = (int) asmsg->body[10];
	      wi->width = asmsg->body[5];
	      wi->height = asmsg->body[6];
	      if (!wins_tail)
		wins_tail = wi;
	      insert_window_list (&wins, wi);
	      ++wins_count;
	    }
	  last = 1;
	  break;

	case M_END_WINDOWLIST:
	  break;

	default:
	  fprintf (stderr,
		   "%s: internal inconsistency: unexpected message\n",
		   MyName);
	  break;
	}
      DestroyASMessage (asmsg);
    }
  return last;
}

void
wait_configure (window_item * wi)
{
  int found = 0;
  unsigned long header[HEADER_SIZE], *body;
  int count;
  fd_set infds;
  FD_ZERO (&infds);
  FD_SET (fd[1], &infds);
  select (fd_width, &infds, 0, 0, NULL);
  while (!found)
    if ((count = ReadASPacket (fd[1], header, &body)) > 0)
      {
	if ((header[1] == M_CONFIGURE_WINDOW)
	    && (Window) body[1] == wi->frame)
	  found = 1;
	free (body);
      }
}

int
atopixel (char *s, unsigned long f)
{
  int l = strlen (s);
  if (l < 1)
    return 0;
  if (isalpha (s[l - 1]))
    {
      char s2[24];
      strcpy (s2, s);
      s2[strlen (s2) - 1] = 0;
      return atoi (s2);
    }
  return (atoi (s) * f) / 100;
}

void
cascade_windows (void)
{
  char msg[128];
  int cur_x = ofsx, cur_y = ofsy;
  window_item *w = reversed ? wins_tail : wins;
  while (w)
    {
      unsigned long nw = 0, nh = 0;
      if (raise_window)
	SendInfo (fd, "Raise", w->frame);
      sprintf (msg, "Move %up %up", cur_x, cur_y);
      SendInfo (fd, msg, w->frame);
      if (resize)
	{
	  if (nostretch)
	    {
	      if (maxw
		  && (w->width > maxw))
		nw = maxw;
	      if (maxh
		  && (w->height > maxh))
		nh = maxh;
	    }
	  else
	    {
	      nw = maxw;
	      nh = maxh;
	    }
	  if (nw || nh)
	    {
	      sprintf (msg, "Resize %lup %lup",
		       nw ? nw : w->width,
		       nh ? nh : w->height);
	      SendInfo (fd, msg, w->frame);
	    }
	}
      wait_configure (w);
      if (!flatx)
	cur_x += w->bw;
      cur_x += incx;
      if (!flaty)
	cur_y += w->bw + w->th;
      cur_y += incy;
      w = reversed ? w->prev : w->next;
    }
}

void
parse_args (char *s, int argc, char *argv[], int argi)
{
  int nsargc = 0;
  /* parse args */
  for (; argi < argc; ++argi)
    {
      if (!strcmp (argv[argi], "-u"))
	{
	  untitled = 1;
	}
      else if (!strcmp (argv[argi], "-t"))
	{
	  transients = 1;
	}
      else if (!strcmp (argv[argi], "-a"))
	{
	  all = untitled = sticky = transients = maximized = 1;
	}
      else if (!strcmp (argv[argi], "-r"))
	{
	  reversed = 1;
	}
      else if (!strcmp (argv[argi], "-noraise"))
	{
	  raise_window = 0;
	}
      else if (!strcmp (argv[argi], "-desk"))
	{
	  desk = 1;
	}
      else if (!strcmp (argv[argi], "-flatx"))
	{
	  flatx = 1;
	}
      else if (!strcmp (argv[argi], "-flaty"))
	{
	  flaty = 1;
	}
      else if (!strcmp (argv[argi], "-m"))
	{
	  maximized = 1;
	}
      else if (!strcmp (argv[argi], "-s"))
	{
	  sticky = 1;
	}
      else if (!strcmp (argv[argi], "-resize"))
	{
	  resize = 1;
	}
      else if (!strcmp (argv[argi], "-nostretch"))
	{
	  nostretch = 1;
	}
      else if (!strcmp (argv[argi], "-incx") && ((argi + 1) < argc))
	{
	  incx = atopixel (argv[++argi], Scr.MyDisplayWidth);
	}
      else if (!strcmp (argv[argi], "-incy") && ((argi + 1) < argc))
	{
	  incy = atopixel (argv[++argi], Scr.MyDisplayHeight);
	}
      else
	{
	  if (++nsargc > 4)
	    {
	      fprintf (stderr,
		       "%s: %s: ignoring unknown arg %s\n",
		       MyName, s, argv[argi]);
	      continue;
	    }
	  if (nsargc == 1)
	    {
	      ofsx = atopixel (argv[argi], Scr.MyDisplayWidth);
	    }
	  else if (nsargc == 2)
	    {
	      ofsy = atopixel (argv[argi], Scr.MyDisplayHeight);
	    }
	  else if (nsargc == 3)
	    {
	      maxw = atopixel (argv[argi], Scr.MyDisplayWidth);
	    }
	  else if (nsargc == 4)
	    {
	      maxh = atopixel (argv[argi], Scr.MyDisplayHeight);
	    }
	}
    }
}

int
parse_line (char *s, char ***args)
{
  int count = 0, i = 0;
  char *arg_save[48];
  strtok (s, " ");
  while ((s = strtok (NULL, " ")))
    arg_save[count++] = s;
  *args = (char **) safemalloc (sizeof (char *) * count);
  for (; i < count; ++i)
    (*args)[i] = arg_save[i];
  return count;
}

char *
GetConfigLine (const char *filename)
{
  FILE *f;
  char *line, *tline;
  int found = 0, len;

  if ((f = fopen (filename, "r")) != NULL)
    {
      line = (char *) safemalloc (MAXLINELENGTH);
      len = strlen (MyName);
      while (((tline = fgets (line, MAXLINELENGTH, f)) != NULL) && (!found))
	{
	  if ((*line == '*') && (!strncmp (line + 1, MyName, len)))
	    {
	      found = 1;
	      break;
	    }
	}
      fclose (f);
      if (found)
	{
	  len = strlen (line);
	  tline = (char *) safemalloc (len);
	  strcpy (tline, line);
	  free (line);
	  if (tline[len - 1] == '\n')
	    tline[len - 1] = 0;
	  return (tline);
	}
      free (line);
    }
  return (NULL);
}

void
DeadPipe (int sig)
{
  exit (0);
}

void
usage (void)
{
  printf ("Usage:\n"
	"%s [--version] [--help] [-u] [-t] [-a] [-r] [-m] [-s] [-noraise]\n"
      "%*s [-desk] [-flatx] [-flaty] [-resize] [-nostretch] [-incx value]\n"
	  "%*s [-incy value] [xoffset yoffset maxwidth maxheight]\n"
	  ,MyName, (int) strlen (MyName), "", (int) strlen (MyName), "");
  exit (0);
}

/*
 * error_handler
 * catch X errors, display the message and continue running.
 */
int
error_handler (Display * disp, XErrorEvent * event)
{
  fprintf (stderr, "%s: internal error, error code %d, request code %d, minor code %d.\n",
	 MyName, event->error_code, event->request_code, event->minor_code);
  return 0;
}

int
main (int argc, char *argv[])
{
  int config_line_count;
  char *config_line;
  int i;
  char *global_config_file = NULL;

  /* Save our program name - for error messages */
  SetMyName (argv[0]);

  i = ProcessModuleArgs (argc, argv, &(global_config_file), NULL, NULL, usage);

  /* Dead pipe == AS died */
  signal (SIGPIPE, DeadPipe);
  signal (SIGQUIT, DeadPipe);
  signal (SIGSEGV, DeadPipe);
  signal (SIGTERM, DeadPipe);

  x_fd = ConnectX (ASDefaultScr, display_name, PropertyChangeMask);
  XSetErrorHandler (error_handler);

  /* connect to AfterStep */
  fd_width = GetFdWidth ();
  fd[0] = fd[1] = ConnectAfterStep (mask_reg);

  if (global_config_file == NULL)
    {
      char configfile[255];
      sprintf (configfile, "%s/base.%dbpp", AFTER_DIR, DefaultDepth (dpy, screen));
      global_config_file = (char *) PutHome (configfile);
      if (CheckFile (global_config_file) == -1)
	{
	  free (global_config_file);
	  sprintf (configfile, "%s/base.%dbpp", AFTER_SHAREDIR, DefaultDepth (dpy, screen));
	  global_config_file = PutHome (configfile);
	}
    }
  if ((config_line = GetConfigLine (global_config_file)) != NULL)
    {
      char **args = NULL;
      config_line_count = parse_line (config_line, &args);
      parse_args ("config args",
		  config_line_count, args, 0);
      free (config_line);
      free (args);
    }
  parse_args ("module args", argc, argv, i);

/* avoid interactive placement in fvwm1
 *    if (!ofsx) ++ofsx;
 *      if (!ofsy) ++ofsy;
 */

  SendInfo (fd, "Send_WindowList", 0);
  while (get_window ());
  if (wins_count)
    cascade_windows ();
  free_window_list (&wins);

  return 0;
}
